sdk/nodejs/state/remoteStateReference.ts

```diff
@@ -0,0 +1,59 @@
+import * as pulumi from "@pulumi/pulumi";
+import {S3BackendConfig} from "./s3BackendConfig";
+import {ConsulBackendConfig} from "./consulBackendConfig";
+
+/**
+ * The set of arguments for constructing a RemoteStateReference resource.
+ */
+export interface RemoteStateReferenceArgs {
+    /**
+     * The configuration of the Terraform Remote State backend, including type.
+     */
+    readonly backendConfig: S3BackendConfig | ConsulBackendConfig;
```

Comment:
Since this is the key part and `workspace` is backend-specific already, should we consider just doing

```typescript
export type RemoteStateReferenceArgs = S3BackendConfig | ConsulBackendConfig;
```

and perhaps rename `S3BackendConfig` to `S3RemoteStateReferenceArgs`, etc? It'd slightly simplify the use of the type. (Possibly a bad idea ... just asking ...)

Reply:
We could do this - in fact it was the original design. The reason I decided against it in the end was needing to delete unused keys from the config before it is passed to the backend - it seems simpler to only have to delete the discriminator key than anything in the config.

Reply:
> The reason I decided against it in the end was needing to delete unused keys from the config before it is passed to the backend

I'm not sure I follow why that would be necessary.  At the JavaScript level (the runtime object), the same subsets of properties are supported in either case - right?

Reply:
In JavaScript yes - but ultimately we get a `structpb` in the providers handler, and have to manipulate that. We cannot call `PrepareConfig` or `CoerceValue` on the `go-cty` object representing the config struct with extraneous values in the map we pass in, so values have to be deleted. I've changed it to make everything a top level property however - there are only two fields to delete instead of one.

Reply:
> with extraneous values in the map we pass in

How are extraneous values getting in here?  How does the TypeScript type impact what values are in here?  (all that should matter is what JavaScript object is passed in)

Reply:
We get the properties in the shape they appear in the object here. If config was in an object, there is nothing to do:

```javascript
{
    config: {
        // all keys in here go to the go-cty object
    },
    workspace: "x" // this _must not_ go to the go-cty object
}
```

Whereas if we have all of the config keys in one object, we have to construct a new object with just the keys in the config schema before being able to pass it to the go-cty object.

However, since there is only one additional key in there, it's not a huge problem (and is now implemented).

Reply:
I see what you are saying. 

> However, since there is only one additional key in there, it's not a huge problem (and is now implemented).

Yeah - this seems okay.